src/hevm/src/EVM/SymExec.hs

```diff
@@ -128,64 +136,54 @@ loadSymVM x initStore addr callvalue' calldata' =
     }) & set (env . contracts . at (createAddress ethrunAddress 1))
              (Just (contractWithStore x initStore))
 
--- Interpreter which explores all paths at
--- branching points.
--- returns a list of possible final evm states
+-- | Interpreter which explores all paths at
+-- | branching points.
+-- | returns a list of possible final evm states
 interpret
   :: Fetch.Fetcher
   -> Maybe Integer --max iterations
   -> Stepper a
-  -> StateT VM IO (Either Stepper.Failure [a])
+  -> StateT VM IO [a]
 interpret fetcher maxIter =
   eval . Operational.view
 
   where
     eval
       :: Operational.ProgramView Stepper.Action a
-      -> StateT VM IO (Either Stepper.Failure [a])
+      -> StateT VM IO [a]
 
     eval (Operational.Return x) =
-      pure (Right [x])
+      pure [x]
 
     eval (action Operational.:>>= k) =
       case action of
         Stepper.Exec ->
           exec >>= interpret fetcher maxIter . k
         Stepper.Run ->
           run >>= interpret fetcher maxIter . k
-        Stepper.Option (EVM.PleaseChoosePath continue) ->
-          do vm <- get
-             case maxIter of
-               Just maxiter ->
-                 let pc' = view (state . pc) vm
-                     addr = view (state . contract) vm
-                     iters = view (iterations . at (addr, pc') . non 0) vm
-                 in if num maxiter <= iters then
-                      vm ^?! (cache . path . ix ((addr, pc'), iters - 1)) & \case
-                        -- When we have reached maxIterations, we take the choice that will hopefully
-                        -- lead us out of here.
-                        Known 0 -> State.state (runState (continue 1)) >> interpret fetcher maxIter (k ())
-                        Known 1 -> State.state (runState (continue 0)) >> interpret fetcher maxIter (k ())
-                        n -> error ("I don't see how this could have happened: " <> show n)
-                    else do a <- State.state (runState (continue 0)) >> interpret fetcher maxIter (k ())
-                            put vm
-                            b <- State.state (runState (continue 1)) >> interpret fetcher maxIter (k ())
-                            return $ liftA2 (<>) a b
-               Nothing -> do a <- State.state (runState (continue 0)) >> interpret fetcher maxIter (k ())
-                             put vm
-                             b <- State.state (runState (continue 1)) >> interpret fetcher maxIter (k ())
-                             return $ liftA2 (<>) a b
+        Stepper.Ask (EVM.PleaseChoosePath continue) -> do
+          vm <- get
+          case maxIterationsReached vm maxIter of
+            Nothing -> do a <- interpret fetcher maxIter (Stepper.evm (continue True) >>= k)
+                          put vm
+                          b <- interpret fetcher maxIter (Stepper.evm (continue False) >>= k)
+                          return $ a <> b
```

Comment:
There might be a way, but I'm not sure if it's gonna be easy. I think it would have been a lot simpler if we wouldn't have an open socket with the solver